packages/hardhat-ethers/src/helpers.ts

```diff
@@ -257,10 +254,14 @@ export async function getContractAt(
   address: string,
   signer?: ethers.Signer
 ) {
-  const { Contract } = require("ethers") as typeof ethers;
-
   if (typeof nameOrAbi === "string") {
-    const factory = await getContractFactoryByName(hre, nameOrAbi, signer);
+    const artifact = await hre.artifacts.readArtifact(nameOrAbi);
+    const factory = await getContractFactoryByAbiAndBytecode(
+      hre,
+      artifact.abi,
+      "0x",
+      signer
+    );
```

Comment:
This seems like the only change that this PR needs. What do the rest of the changes do? They seem to inline some code that was in the `getContractFactoryByName` function, so it doesn't really seem like an improvement in readability. Am I missing something here?

Reply:
I just added the `getContractFactoryByName` function back and restructured this into several commits so it's clearer what it is doing.